GH-37093: [Python] Add async Flight client with GetFlightInfo

lidavidm opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

Demonstrate how to deal with async Flight from Python.

### What changes are included in this PR?

Add bindings for GetFlightInfo.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes, new APIs.

* Closes: #37093

lidavidm commented:
CC @pitrou @jorisvandenbossche this is a skeleton of how to bind async C++ code to asyncio in Python, though I still need to get it working with both asyncio and Trio. (I'm looking at https://github.com/codypiersall/pynng/blob/master/pynng/_aio.py as a reference of how we might do that.)

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -1219,6 +1224,80 @@ cdef class FlightMetadataWriter(_Weakrefable):
+        *,
+        options: FlightCallOptions = None,
+    ):
+        call = AsyncCall()
```
Could pass the asyncio loop to `AsyncCall` here.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -1219,6 +1224,80 @@ cdef class FlightMetadataWriter(_Weakrefable):
+    call = <object> self
+    if not status.ok():
+        print("Failed")
+        call.exception = RuntimeError(status.ToString())
```
Should use our regular Status conversion routines.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -1320,6 +1399,11 @@ cdef class FlightClient(_Weakrefable):
 
+    def as_async(self) -> None:
+        if anyio is None:
+            raise RuntimeError("anyio is required for the async interface")
```
I'm not sure why we would want `anyio`. Just start with plain `asyncio`?

lidavidm replied:
The main thing is to support both asyncio and Trio users (though I think anyio is unnecessary and we can just directly support them both)

pitrou replied:
Why would we want that, though? asyncio is the de facto standard. And the anyio API for delegating stuff to external threads seems rather convoluted.

lidavidm replied:
I don't see a reason to just lock out a part of the ecosystem.

lidavidm replied:
And FWIW, one of the very first issues opened against ADBC was request for async support via anyio specifically, to support both Trio and asyncio

pitrou commented on python/pyarrow/src/arrow/python/flight.cc:
```diff
@@ -383,6 +383,31 @@ Status CreateSchemaResult(const std::shared_ptr<arrow::Schema>& schema,
+    const arrow::flight::FlightCallOptions& options,
+    const arrow::flight::FlightDescriptor& descriptor,
+    PyObject* context, AsyncGetFlightInfoCallback callback) {
+  // TODO: OwnedRefNoGil the context?
```
Should certainly do so indeed.

pitrou reviewed (commented):
This looks like a good start to me.

lidavidm commented:
Updated, added basic tests. I'll deal with Trio in a later PR, as well as raising an RpcError-type exception instead of built in exceptions (again: RPC errors should be distinct, and should be handled uniformly), once I've reviewed the C++ side and decided I'm happy with TransportStatusDetail.

lidavidm commented on python/pyarrow/_flight.pyx:
```diff
@@ -1219,6 +1219,94 @@ cdef class FlightMetadataWriter(_Weakrefable):
+            self._exception = exception
+        else:
+            self._result = result
+        # XXX: why does set() have to be done from within the loop?
```
Ah, I need to dig into this still.

lidavidm replied:
Also, I wonder if we can just directly complete an asyncio.Future (though I am not sure if that generalizes to Trio, or to streaming RPCs)

lidavidm commented:
Hmm, that's an interesting segfault in CI.

lidavidm commented:
Alright:

- The weird `SystemError: <class 'SystemError'> returned a result with an error set` is inscrutable to me. It occurs along with a `TypeError: __traceback__ must be a traceback or None` message. `with_traceback` doesn't appear to be able to clear the problematic traceback and I can't get gdb to break on the source of the error properly.
- The segfault is also related to tracebacks, somehow; it reproduces with PYTHONDEVMODE and gdb shows it happens somewhere in `tb_traverse`, where `obj->ob_type` points to something that appears to have already been deallocated:

  ```
  #0  _PyObject_IS_GC (obj=0x7fff656ce7af) at /usr/local/src/conda/python-3.11.3/Include/object.h:772
  #1  visit_decref (op=0x7fff656ce7af, parent=parent@entry=0x7fff656ce710)
      at /usr/local/src/conda/python-3.11.3/Modules/gcmodule.c:452
  #2  0x000055555583df5b in tb_traverse (tb=0x7fff656ce710, visit=0x555555723a20 <visit_decref>, arg=0x7fff656ce710)
      at /usr/local/src/conda/python-3.11.3/Python/traceback.c:182
  ```

lidavidm commented:
So, I don't understand why, but bouncing through a pure-Python thread fixes the segfault. It seems something about having the C-callback in the traceback at some point borks the interpreter. It does seem that the asyncio implementation tries to get the traceback unconditionally (in call_soon_threadsafe -> call_soon -> asyncio.events.Handle) so I suppose maybe Cython isn't giving a valid traceback object for `_client_async_get_flight_info`? (Should probably go confirm that.)

No clue about the first error...

pitrou commented:
Can the segfault be reproduced locally?

pitrou commented:
Ok, I'll take a look hopefully on Monday.

lidavidm commented:
Looking at these:
- https://github.com/Delgan/loguru/issues/88
- https://github.com/cython/cython/issues/2735

It looks like Cython indeed doesn't generate stack frames for calls, and also modifying `wakeup` here to try to call `sys._getframe` like `asyncio.event.Handle` does crashes with `call stack is not deep enough`. So indeed, it seems we should avoid touching asyncio or generating exceptions from the native callback. Bouncing through a Python thread seems unavoidable from that perspective, unfortunately.

lidavidm commented:
gRPC themselves do something quite different; I don't know if they ran into challenges like this, but they sidestep this neatly: https://ep2020.europython.eu/media/conference/slides/AfUnDYW-grpc-python-c-extensions-and-asyncio.pdf

> - gRPC Asyncio Python application would start a separated thread for polling gRPC events
> - This thread won’t use any Python object, during the polling
> - Events would be added into a C++ queue
> - Asyncio loop will be woken up by writing into a socket


(Note that grpcio in Python works quite differently and implements a lot of what gRPC-C++ provides itself, so that may also explain the architecture. In particular, they have to drive the gRPC event loop, while gRPC-C++ manages that for us.)

For now, though, I think the easy way would be to have a single-thread thread pool per client that's used to trigger these callbacks.

lidavidm commented:
Our callback never raises an exception. However it may generate an exception that we raise from within the asyncio loop. (I guess asyncio must attach the 'bad' traceback to it somewhere, as context for debugging? Since if we don't generate an exception, things are fine.)

pitrou commented:
> Our callback never raises an exception.

It does not do so deliberately, sure, but it would raise one unwillingly?

lidavidm commented:
Hmm, maybe there is one and it's getting suppressed, or rather, Python trying to print the exception is what's crashing the interpreter?

<details>

```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
PyFrame_GetCode (frame=frame@entry=0x7fff6a05257000) at /usr/local/src/conda/python-3.11.3/Objects/frameobject.c:1312
1312	/usr/local/src/conda/python-3.11.3/Objects/frameobject.c: No such file or directory.
(gdb) bt
#0  PyFrame_GetCode (frame=frame@entry=0x7fff6a05257000) at /usr/local/src/conda/python-3.11.3/Objects/frameobject.c:1312
#1  0x000055555569728e in tb_printinternal (margin=<optimized out>, indent=<optimized out>, limit=<optimized out>, 
    f=<optimized out>, tb=0x7ffff0ba8aef) at /usr/local/src/conda/python-3.11.3/Python/traceback.c:943
#2  _PyTraceBack_Print_Indented (v=v@entry=0x7fff6a3b0ff0, indent=0, margin=0x55555588988a "", 
    header_margin=<optimized out>, header=<optimized out>, f=0x7ffff705d8b0)
    at /usr/local/src/conda/python-3.11.3/Python/traceback.c:1018
#3  0x000055555568b9dc in print_exception_traceback (value=0x7fff684a9fd0, ctx=0x7fffffffd258)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:942
#4  print_exception (ctx=ctx@entry=0x7fffffffd258, value=value@entry=0x7fff684a9fd0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1224
#5  0x000055555568b233 in print_exception_recursive (ctx=ctx@entry=0x7fffffffd258, value=value@entry=0x7fff684a9fd0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1509
#6  0x000055555568b054 in _PyErr_Display (file=0x7ffff705d8b0, exception=<optimized out>, value=0x7fff684a9fd0, 
    tb=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1552
#7  0x000055555568acf1 in PyErr_Display (exception=0x555556230ba0, value=0x7fff684a9fd0, tb=0x7fff6a3b0ff0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1584
#8  0x000055555568abe7 in sys_excepthook_impl (traceback=<optimized out>, value=<optimized out>, 
    exctype=<optimized out>, module=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/sysmodule.c:745
#9  sys_excepthook (module=<optimized out>, args=0x7fffffffd380, nargs=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Python/clinic/sysmodule.c.h:73
#10 0x000055555575785b in cfunction_vectorcall_FASTCALL (func=func@entry=0x7ffff7250050, args=0x7fffffffd380, nargsf=3, 
    kwnames=<optimized out>) at /usr/local/src/conda/python-3.11.3/Objects/methodobject.c:427
#11 0x00005555557299b5 in _PyObject_VectorcallTstate (tstate=0x555555ad08b8 <_PyRuntime+166328>, 
    callable=0x7ffff7250050, args=<optimized out>, nargsf=<optimized out>, kwnames=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Include/internal/pycore_call.h:92
#12 0x0000555555828eae in _PyErr_PrintEx (tstate=0x555555ad08b8 <_PyRuntime+166328>, set_sys_last_vars=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:826
#13 0x000055555568a801 in PyErr_Print () at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:882
#14 _PyRun_SimpleFileObject (fp=<optimized out>, filename=<optimized out>, closeit=<optimized out>, 
    flags=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:446
#15 0x0000555555828a23 in _PyRun_AnyFileObject (fp=0x555555b225b0, filename=0x7ffff72672e0, closeit=1, 
    flags=0x7fffffffd498) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:79
#16 0x0000555555822ab6 in pymain_run_file_obj (skip_source_first_line=0, filename=0x7ffff72672e0, 
    program_name=0x7ffff71e7040) at /usr/local/src/conda/python-3.11.3/Modules/main.c:360
#17 pymain_run_file (config=0x555555ab6900 <_PyRuntime+59904>) at /usr/local/src/conda/python-3.11.3/Modules/main.c:379
#18 pymain_run_python (exitcode=0x7fffffffd490) at /usr/local/src/conda/python-3.11.3/Modules/main.c:601
#19 Py_RunMain () at /usr/local/src/conda/python-3.11.3/Modules/main.c:680
#20 0x00005555557e5179 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Modules/main.c:734
#21 0x00007ffff7c58083 in __libc_start_main (main=0x5555557e50d0 <main>, argc=2, argv=0x7fffffffd6c8, 
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd6b8)
    at ../csu/libc-start.c:308
#22 0x00005555557e5015 in _start () at /usr/local/src/conda/python-3.11.3/Python/getplatform.c:9
```

</details>

lidavidm commented:
Ah, _and_ the crash is now non-deterministic...

pitrou commented on python/pyarrow/src/arrow/python/flight.cc:
```diff
@@ -383,6 +383,25 @@ Status CreateSchemaResult(const std::shared_ptr<arrow::Schema>& schema,
+      } else {
+        callback(py_context.obj(), nullptr, result.status());
+      }
+      return Status::OK();
```
Perhaps this instead, to catch and bubble up the exception neatly?
```suggestion
      return CheckPyError();
```

lidavidm replied:
I do have this locally.

lidavidm replied:
I pushed that up

lidavidm commented:
```
Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
```

Hmm, am I doing something wrong on the C++ side?

lidavidm commented:
Hmm, yeah.

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f25576b8859 in __GI_abort () at abort.c:79
#2  0x0000558134e9cf78 in fatal_error_exit (status=-1) at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2722
#3  fatal_error (fd=<optimized out>, header=header@entry=1, 
    prefix=prefix@entry=0x5581350bb890 <__func__.5.lto_priv.1> "_PyMem_DebugFree", 
    msg=msg@entry=0x55813504c2f0 "Python memory allocator called without holding the GIL", status=status@entry=-1)
    at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2833
#4  0x0000558134e9d334 in _Py_FatalErrorFunc (func=func@entry=0x5581350bb890 <__func__.5.lto_priv.1> "_PyMem_DebugFree", 
    msg=msg@entry=0x55813504c2f0 "Python memory allocator called without holding the GIL")
    at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2919
#5  0x0000558134e02b3c in _PyMem_DebugCheckGIL (func=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Objects/obmalloc.c:2683
#6  _PyMem_DebugFree (ctx=<optimized out>, ptr=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Objects/obmalloc.c:2707
#7  0x0000558134fe37f5 in _PyFaulthandler_Fini () at /usr/local/src/conda/python-3.11.3/Modules/faulthandler.c:1445
#8  0x0000558134e9d2ca in fatal_error (fd=2, header=header@entry=1, 
    prefix=prefix@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get", 
    msg=msg@entry=0x55813504ddc0 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)", status=status@entry=-1) at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2891
#9  0x0000558134e9d334 in _Py_FatalErrorFunc (
    func=func@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get", 
    msg=msg@entry=0x55813504ddc0 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)") at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2919
#10 0x0000558134e9d3e4 in _Py_FatalError_TstateNULL (
    func=func@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get")
    at /usr/local/src/conda/python-3.11.3/Python/ceval.c:344
#11 0x0000558134e31174 in _Py_EnsureFuncTstateNotNULL (tstate=<optimized out>, func=<synthetic pointer>)
    at /usr/local/src/conda/python-3.11.3/Include/internal/pycore_pystate.h:94
#12 PyThreadState_Get () at /usr/local/src/conda/python-3.11.3/Python/pystate.c:1200
#13 subtype_dealloc (self=0x7f24c9ae4dd0) at /usr/local/src/conda/python-3.11.3/Objects/typeobject.c:1369
#14 0x00007f24c97aad52 in Py_DECREF (op=0x7f24c9ae4dd0)
    at /home/lidavidm/miniforge3/envs/dev/include/python3.11/object.h:538
#15 0x00007f24c97aad74 in Py_XDECREF (op=0x7f24c9ae4dd0)
    at /home/lidavidm/miniforge3/envs/dev/include/python3.11/object.h:602
#16 0x00007f24c97ab2e7 in arrow::py::OwnedRef::reset (this=0x55813774ba48, obj=0x0)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/common.h:150
#17 0x00007f24c97ab313 in arrow::py::OwnedRef::reset (this=0x55813774ba48)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/common.h:154
#18 0x00007f24c97a8e18 in operator() (__closure=0x55813774ba28, result=...)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/flight.cc:403
#19 0x00007f24c97aa7b4 in arrow::Future<arrow::flight::FlightInfo>::WrapResultyOnComplete::Callback<arrow::py::flight::AsyncGetFlightInfo(arrow::flight::FlightClient*, const arrow::flight::FlightCallOptions&, const arrow::flight::FlightDescriptor&, PyObject*, arrow::py::flight::AsyncGetFlightInfoCallback)::<lambda(arrow::Result<arrow::flight::FlightInfo>)> >::operator()(const arrow::FutureImpl &) (this=0x55813774ba28, impl=...)
    at /home/lidavidm/Code/upstream/flight-async/install/include/arrow/util/future.h:442
```

pitrou commented:
You probably need to turn the `OwnedRef` into an `OwnedRefNoGIL`?


Le 10/08/2023 à 19:34, David Li a écrit :
> Hmm, yeah.

lidavidm commented:
It *is* an OwnedRefNoGIL though (not sure why the backtrace is like that, presumably because release() is defined on the base class)

lidavidm commented:
Ok, well setting that aside (just taking out the explicit release() seems to be fine), the issue is still that somehow, we're generating an exception with an invalid traceback, and when we either go to print that traceback out or garbage collect it, the interpreter dies. I wonder if it has something to do with us generating the exception from a cdef function + asyncio possibly doing something weird/unsafe with stackframes that doesn't mesh well with Cython cdef functions.

lidavidm commented:
And then if I replace check_status with just a hardcoded Python exception, I still get a traceback with an invalid frame being garbage-collected somewhere.

lidavidm commented:
Ok. I think I figured it out.

I forgot to Py_INCREF before using the OwnedRefNoGIL :facepalm:

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -1219,6 +1220,93 @@ cdef class FlightMetadataWriter(_Weakrefable):
+        result.info.reset(new CFlightInfo(move(deref(info))))
+        call.wakeup(result=result)
+    else:
+        call.wakeup(exception=convert_status(status))
```
I've sketched something that seems to work here:
https://gist.github.com/pitrou/d4e5a386bc62d05fda6593f1b1e83a47

lidavidm replied:
Thanks! Applied the patch + added an async.h as noted

lidavidm commented on python/pyarrow/_flight.pyx:
```diff
@@ -1320,6 +1410,11 @@ cdef class FlightClient(_Weakrefable):
+    def as_async(self) -> None:
+        if self._pool is None:
+            self._pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        return AsyncioFlightClient(self, self._pool)
```
It's not, I forgot to remove this after trying something

lidavidm commented:
Yes, this should be ready.

pitrou reviewed (approved):
Thanks a lot. This really opens interesting possibilities, I think.

pitrou commented:
@github-actions crossbow submit -g python

pitrou commented:
At least `test-debian-11-python-3` has failed with:
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=53319&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181&l=5448
```
usr/local/lib/python3.9/dist-packages/pyarrow/tests/test_flight_async.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/lib/python3.9/asyncio/runners.py:44: in run
    return loop.run_until_complete(main)
usr/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
    return future.result()
usr/local/lib/python3.9/dist-packages/pyarrow/tests/test_flight_async.py:56: in _test
    info = await async_client.get_flight_info(descriptor)
pyarrow/_flight.pyx:1281: in get_flight_info
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pyarrow.lib.ArrowNotImplementedError: gRPC 1.40 or newer is required to use async
```

We could perhaps expose `supports_async()` in Python.

pitrou commented:
That said, it would perhaps be nicer to expose the same error message in Python as in C++...

pitrou commented:
@github-actions crossbow submit -g python

pitrou commented on cpp/src/arrow/flight/client.h:
```diff
@@ -396,6 +396,13 @@ class ARROW_FLIGHT_EXPORT FlightClient {
+  /// This is like supports_async(), except that a detailed error message
+  /// is returned if async support is not available.  If async support is
+  /// available, this function returns successfully.
+  Status CheckAsyncSupport() const;
```
@lidavidm Does this API look ok?

pitrou commented:
CI failures are unrelated.
